core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/OClusterPage.java

```diff
@@ -481,39 +476,70 @@ private boolean isPositionInsideInterval(final int recordPosition) {
+    final int recordsCount = getRecordsCount();
+    final int freePosition = getIntValue(FREE_POSITION_OFFSET);
 
-    int currentPosition = freePosition;
-    final List<Integer> processedPositions = new ArrayList<Integer>();
+    // 1. Build the entries "map" and merge consecutive holes.
 
+    final int maxEntries = recordsCount /* live records */ + recordsCount + 1 /* max holes after merging */;
+    final int[] positions = new int[maxEntries];
+    final int[] sizes = new int[maxEntries];
+
+    int count = 0;
+    int currentPosition = freePosition;
+    int lastEntryKind = ENTRY_KIND_UNKNOWN;
     while (currentPosition < PAGE_SIZE) {
-      int entrySize = getIntValue(currentPosition);
+      final int size = getIntValue(currentPosition);
+      final int entryKind = Integer.signum(size);
+      assert entryKind != ENTRY_KIND_UNKNOWN;
 
-      if (entrySize > 0) {
-        int positionIndex = getIntValue(currentPosition + OIntegerSerializer.INT_SIZE);
-        processedPositions.add(positionIndex);
+      if (entryKind == ENTRY_KIND_HOLE && lastEntryKind == ENTRY_KIND_HOLE)
+        sizes[count - 1] += size;
+      else {
+        positions[count] = currentPosition;
+        sizes[count] = size;
 
-        currentPosition += entrySize;
-      } else {
-        entrySize = -entrySize;
-        moveData(freePosition, freePosition + entrySize, currentPosition - freePosition);
-        currentPosition += entrySize;
-        freePosition += entrySize;
+        ++count;
 
-        shiftPositions(processedPositions, entrySize);
+        lastEntryKind = entryKind;
       }
+
+      currentPosition += entryKind == ENTRY_KIND_HOLE ? -size : size;
     }
 
-    setIntValue(FREE_POSITION_OFFSET, freePosition);
-  }
+    // 2. Iterate entries in reverse, update data offsets, merge consecutive data segments and move them in a single operation.
+
+    int shift = 0;
+    int lastDataPosition = 0;
+    int mergedDataSize = 0;
+    for (int i = count - 1; i >= 0; --i) {
+      final int position = positions[i];
+      final int size = sizes[i];
 
-  private void shiftPositions(final List<Integer> processedPositions, final int entrySize) throws IOException {
-    for (int positionIndex : processedPositions) {
-      final int entryIndexPosition = PAGE_INDEXES_OFFSET + INDEX_ITEM_SIZE * positionIndex;
-      final int entryPosition = getIntValue(entryIndexPosition);
-      setIntValue(entryIndexPosition, entryPosition + entrySize);
+      final int entryKind = Integer.signum(size);
+      assert entryKind != ENTRY_KIND_UNKNOWN;
+
+      if (entryKind == ENTRY_KIND_DATA && shift > 0) {
+        final int positionIndex = getIntValue(position + OIntegerSerializer.INT_SIZE);
+        setIntValue(PAGE_INDEXES_OFFSET + INDEX_ITEM_SIZE * positionIndex, position + shift);
+
+        lastDataPosition = position;
+        mergedDataSize += size;
+      }
+
+      if (mergedDataSize > 0 && (entryKind == ENTRY_KIND_HOLE || i == 0)) {
```

Comment:
Please add a comment for this branch because it is not clear at first glance why we need to accumulate size of the data.